presentation-compiler/src/main/dotty/tools/pc/completions/Completions.scala

```diff
@@ -228,33 +226,47 @@ class Completions(
   def completionsWithSuffix(
       denot: SingleDenotation,
       label: String,
-      toCompletionValue: (String, SingleDenotation, CompletionSuffix) => CompletionValue
+      toCompletionValue: (String, SingleDenotation, CompletionAffix) => CompletionValue
   ): List[CompletionValue] =
     val sym = denot.symbol
-    // find the apply completion that would need a snippet
+    val hasNonSyntheticConstructor = sym.name.isTypeName && sym.isClass
+      && !sym.is(ModuleClass) && !sym.is(Trait) && !sym.is(Abstract) && !sym.is(Flags.JavaDefined)
+
     val methodDenots: List[SingleDenotation] =
-      if shouldAddSnippet && completionMode.is(Mode.Term) &&
-        (sym.is(Flags.Module) || sym.isField || sym.isClass && !sym.is(Flags.Trait)) && !sym.is(Flags.JavaDefined)
-      then
-        val info =
-          /* Companion will be added even for normal classes now,
-           * but it will not show up from classpath. We can suggest
-           * constructors based on those synthetic applies.
-           */
-          if sym.isClass && sym.companionModule.exists then sym.companionModule.info
-          else denot.info
-        val applyDenots = info.member(nme.apply).allSymbols.map(_.asSeenFrom(info).asSingleDenotation)
-        denot :: applyDenots
+      if shouldAddSnippet && isNew && hasNonSyntheticConstructor then
+        sym.info.member(nme.CONSTRUCTOR).allSymbols.map(_.asSingleDenotation)
+          .filter(_.symbol.isAccessibleFrom(denot.info))
+      else if shouldAddSnippet && completionMode.is(Mode.Term) && sym.name.isTermName && !sym.is(Flags.Method) && !sym.is(Flags.JavaDefined) then
+        val constructors = if sym.isAllOf(ConstructorProxyModule) then
+          sym.companionClass.info.member(nme.CONSTRUCTOR).allSymbols
+        else
+          val companionApplies  = denot.info.member(nme.apply).allSymbols
+          val classConstructors = if sym.companionClass.exists then
+            sym.companionClass.info.member(nme.CONSTRUCTOR).allSymbols
+          else Nil
+
+          if companionApplies.exists(_.is(Synthetic)) then
+            companionApplies ++ classConstructors.filter(!_.isPrimaryConstructor)
+          else
+            companionApplies ++ classConstructors
+
+        val extraDenots = constructors.map(_.asSeenFrom(denot.info).asSingleDenotation)
+          .filter(_.symbol.isAccessibleFrom(denot.info))
+
+        if sym.isAllOf(ConstructorProxyModule) || sym.is(Trait) then extraDenots
+        else denot :: extraDenots
+
       else denot :: Nil
 
+    val requiresInitDisambiguiation = methodDenots.exists(_.symbol.isConstructor) && methodDenots.exists(_.symbol.name == nme.apply)
```

Comment:
Nah, there is a single case when we don't want to add new.

https://scastie.scala-lang.org/MzTJe12qRkGdcnEhma868g

Reply:
Actually this requires us to add new, so my check will fail anyway :)

Reply:
I only meant that this `methodDenots.exists(_.symbol.isConstructor)` part is excessive since you use it then like so: `methodDenot.symbol.isConstructor && requiresInitDisambiguiation`.

Reply:
Right.